### PR TITLE
feat(trading-competition): Trading competition

### PR DESCRIPTION
**Pull request by ChefHutch**
## To be released Monday 5th April @ 9:00 GMT / 16:00 SGT

**When reviewing, the most notable components would be:**
All within `...Views/TradingCompetition`:

- `TradingCompetition/index` (fetching lots from smart contract, rendering page)
- `components/Countdown` (responsible for timer)
- `components/BattleCta` (responsible for lots of conditional logic with the CTA & subsequent modals)
- `components/RegisterModal/RegisterWithProfile` (responsible for triggering register contract call)

Claim stuff is also in here and significant (`components/ClaimModal`), and tested with Nyan - but it needs UI work which can happen after launch. 

**Needed for registration phase:**
- `BattleCta & smart contracts` : https://github.com/ChefHutch/pancake-frontend/pull/4 ✅  Merged
- `Countdown & competition timing / state` : https://github.com/ChefHutch/pancake-frontend/pull/5 ✅  Merged

**Needed for competition phase - not needed for this PR:**
- `YourScore`: https://github.com/ChefHutch/pancake-frontend/pull/2 [WIP]

To do
- [ ] Make background colours for sections respond to dark mode theme
- [ ] Connect `isCompetitionLive` & `hasCompetitionFinished` conditions to `CompetitionCountdownContext`

### CODE REVIEW

File: src/views/TradingCompetition/components/CompetitionHeadingText.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { Heading } from '@pancakeswap-libs/uikit'
+import styled from 'styled-components'
+
+const sharedStyles = `
+font-size: 32px;
```
Comment (ChefHutch) [inline src/views/TradingCompetition/components/CompetitionHeadingText.tsx:5]: There are a few things about the heading text on the Competition page that differ from our uikit

1. The H1s are 64px on Desktop, 44px on Mobile
2. The H2s are 40px on Desktop, 32px on Mobile
3. They all make use of some gradient stroke stiles. 

I'll manage all of this in the FE for now and then we can see about how to translate them over to the uikit & our existing Headings

Screenshot 2021-03-19 at 08 09 54

File: src/views/TradingCompetition/components/Sticker.tsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import styled from 'styled-components'
+import { Flex } from '@pancakeswap-libs/uikit'
+
+interface StickerProps {
+  backgroundColor?: string
+  borderColor?: string
+}
+
+const Sticker = styled(Flex)<StickerProps>`
```
Comment (ChefHutch) [inline src/views/TradingCompetition/components/Sticker.tsx:9]: This is just a circular flex with a border & bg colour to wrap around circular images that need a 'sticker' border and drop shadow around them, i.e.

Screenshot 2021-03-19 at 16 32 00

Screenshot 2021-03-19 at 16 32 13

File: src/views/TradingCompetition/components/StyledLaurel.tsx [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import styled from 'styled-components'
+import { Laurel } from '../svgs'
```
Comment (ChefHutch) [inline src/views/TradingCompetition/components/StyledLaurel.tsx:2]: The Laurel has been added to the UI kit - it'll be imported from there eventually, just have it here for working with.

File: src/views/TradingCompetition/components/Section/index.tsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+  padding-top: ${({ intersectionPosition }) => (intersectionPosition === 'top' ? '6px' : '48px')};
+  margin: ${({ intersectionPosition }) => (intersectionPosition === 'top' ? '0' : '-34px')} 0;
+
+  ${({ theme }) => theme.mediaQueries.sm} {
+    margin: ${({ intersectionPosition }) => (intersectionPosition === 'top' ? '0' : '-42px')} 0;
+  }
+`
+
+const ChildrenWrapper = styled.div`
```
Comment (ChefHutch) [inline src/views/TradingCompetition/components/Section/index.tsx:21]: This component essentially applies the `Page` (and by extension `Container`) styles minus the min-height to the sections' `ChildrenWrapper`.

The min-height is applied to the `TradingCompetition` index wrapper.

Not sure if there's a more 'scalable' solution here, or if you're happy with this?

File: src/views/TradingCompetition/components/CompetitionHeadingText.tsx (hunk 1 above)
Comment (RabbitDoge) [reply src/views/TradingCompetition/components/CompetitionHeadingText.tsx:15]: See with Bugs if it's really needed. It's better to stick to the uikit rules as much as we can

File: src/views/TradingCompetition/components/CompetitionHeadingText.tsx [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+white-space: nowrap;
+
+${theme.mediaQueries.sm} {
+  font-size: 40px
+}
+`
+
+const sharedStyles = (props) => `
+color: ${props.textColor ? props.textColor : '#ffffff'};
+background:  ${props.background ? props.background : '-webkit-linear-gradient(#7645d9 0%, #452a7a 100%)'};
```
Comment (RabbitDoge) [inline src/views/TradingCompetition/components/CompetitionHeadingText.tsx:26]: Styled component add the vendor prefixes for you. `linear-gradient` is probably enough. Same below

File: src/views/TradingCompetition/components/CompetitionHeadingText.tsx [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  props.fill
+    ? `-webkit-text-fill-color: transparent;`
+    : `
+-webkit-text-stroke: 4px transparent;
+text-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);`
+}
+`
+
+const sharedVisiblyHiddenStyles = `
+visibility: hidden;
```
Comment (RabbitDoge) [inline src/views/TradingCompetition/components/CompetitionHeadingText.tsx:39]: is this prettified ? We should use the same indentation level than in styled components

File: src/views/TradingCompetition/components/Ribbon/index.tsx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+import { Heading2Text, VisuallyHiddenHeading2Text } from '../CompetitionHeadingText'
+
+const Wrapper = styled(Flex)<{ ribbonDirection?: 'up' | 'down' }>`
+  position: relative;
+  display: inline-flex;
+  align-items: ${({ ribbonDirection }) => (ribbonDirection === 'up' ? 'flex-end' : 'flex-start')};
+  justify-content: center;
+
+  svg {
+    z-index: 200;
```
Comment (RabbitDoge) [inline src/views/TradingCompetition/components/Ribbon/index.tsx:22]: Add a comment for why the z-index need to be changed, and why the value 200 is picked

File: src/views/TradingCompetition/components/Ribbon/index.tsx [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+const LaurelWrapper = styled.div<{ dir?: 'l' | 'r' }>`
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  position: absolute;
+  z-index: 2;
+  top: 50%;
+  transform: translate(0, -50%);
+  ${({ dir }) =>
+    dir === 'l'
```
Comment (RabbitDoge) [inline src/views/TradingCompetition/components/Ribbon/index.tsx:43]: wtf is this formatting 😂

File: src/views/TradingCompetition/components/Section/index.tsx (hunk 4 above)
Comment (RabbitDoge) [reply src/views/TradingCompetition/components/Section/index.tsx:21]: You wrote it yourself, it uses the Page/Container style, then why would you duplicate the content here

File: src/views/TradingCompetition/components/Sticker.tsx [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  width: fit-content;
+  height: fit-content;
+  background-color: ${({ theme, backgroundColor }) => backgroundColor || theme.colors.invertedContrast};
+  border: 2px solid ${({ theme, borderColor }) => borderColor || theme.colors.invertedContrast};
+  border-radius: 50%;
```
Comment (RabbitDoge) [inline src/views/TradingCompetition/components/Sticker.tsx:14]: theme.radii.circle

**Review by RabbitDoge — COMMENTED** [review]
I saw you used a lot of hardcoded colors. Make sure there is no mistakes, and check if these colors are not already in the theme. If tthey are new addition to our palette, maybe we need to update the theme files ?

File: src/views/TradingCompetition/components/Section/index.tsx (hunk 4 above)
Comment (ChefHutch) [reply src/views/TradingCompetition/components/Section/index.tsx:21]: Yea man I can just overwrite the Page style that I don't want. My brain must have been much when I did this initially

File: src/views/TradingCompetition/components/CompetitionHeadingText.tsx (hunk 5 above)
Comment (ChefHutch) [reply src/views/TradingCompetition/components/CompetitionHeadingText.tsx:26]: Ah yea you're right on this one. Weirdly some of them do need a `-webkit-` prefix to work (background-clip, text-fill-color, text-stroke)

File: src/views/TradingCompetition/components/CompetitionHeadingText.tsx (hunk 6 above)
Comment (ChefHutch) [reply src/views/TradingCompetition/components/CompetitionHeadingText.tsx:39]: Yea nice - prettier is happy with this junk styling. I've manually set the indents 👍

File: src/views/TradingCompetition/components/Ribbon/index.tsx (hunk 8 above)
Comment (ChefHutch) [reply src/views/TradingCompetition/components/Ribbon/index.tsx:43]: OH this? I added this to make your eyes bleed.

Comment (ChefHutch) [reply src/views/TradingCompetition/components/Ribbon/index.tsx:43]: I actually think this was how linting on my old work laptop was setup. Absolutely brutal eh.

**Comment by ChefHutch** [comment]
> I saw you used a lot of hardcoded colors. Make sure there is no mistakes, and check if these colors are not already in the theme. If tthey are new addition to our palette, maybe we need to update the theme files ?

@RabbitDoge yea - they're pretty much all instances where the colors do exist in the theme, but they aren't meant to change with the light/dark switcher. Will be sure to check with Bugs tho

File: src/views/TradingCompetition/components/Ribbon/index.tsx (hunk 7 above)
Comment (ChefHutch) [reply src/views/TradingCompetition/components/Ribbon/index.tsx:22]: I think this was from me debugging some ribbon shiii. Gonna 🗑️ 👍

File: src/views/TradingCompetition/components/PrizesInfo/index.tsx [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import React from 'react'
+import styled from 'styled-components'
+import { Flex } from '@pancakeswap-libs/uikit'
+import PrizesText from './PrizesText'
+import PrizesTable from './PrizesTable'
+
+const Wrapper = styled(Flex)`
+  max-width: 736px;
```
Comment (ChefHutch) [inline src/views/TradingCompetition/components/PrizesInfo/index.tsx:8]: You'll see this value in a few places for these central 'info' sections. Was a bit of a yolo value picked by bugs on a whim. I do have the standard 1200px set somewhere else, but I guess he wanted it pulling in more?

File: src/App.tsx [hunk 11]
```diff
@@ -41,7 +42,7 @@ const App: React.FC = () => {
   }, [])
 
   useEagerConnect()
-  useFetchPublicData()
+  // useFetchPublicData()
```
Comment (ChefHutch) [inline src/App.tsx:45]: Needed for testnet to function. Once contract is deployed on main net I will remove

File: src/views/TradingCompetition/components/RegisterModal/RegisterWithProfile.tsx [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+  const tradingCompetitionContract = useTradingCompetitionContract()
+  const { account } = useWeb3React()
+  const { toastSuccess, toastError } = useToast()
+  const TranslateString = useI18n()
+
+  const handleConfirmClick = () => {
+    tradingCompetitionContract.methods
+      .register()
+      .send({ from: account })
+      .on('sending', () => {
```
Comment (hachiojidev) [inline src/views/TradingCompetition/components/RegisterModal/RegisterWithProfile.tsx:30]: Just a note. I found out that technically `on` doesn't have the `sending` event although it still seems to work.

`once` does however.

**Review by hachiojidev — APPROVED** [review]

**Pull request merged**